### DISCUSSION TITLE
Elsa 2.1 ✨

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#1268] [Design & tech]
The 2.1 release turned out to be much bigger than anticipated and contains a many a great enhancements and fixes, but also includes a number of **breaking changes**, so please take caution when upgrading your production environment and **create backups**!

### What's New

* Add support for dynamic outcome provider - Closes #873 - 16c93e15c534706615ea2b235535bec847549e01
* Implement support for activity property ordering - Closes #884 - 0f8a0e9f3d9b113b6be7d9da5057f3c1434c405c
* Add Conductor project and SendCommand activity - 589cab520b1e23a88bbfac5ed96b992e2a6b8166
* Add State activity to model finite state machines - 24e378ebfc2207da817292a3008048f23a834b9c
* Add capability to use MessageModel as the input of the Azure Service Bus Sender - f28c9a8be534fdb7c4ad67ad81f6f56d70b85ac5
* Add radio list control - Closes #1062 - c3df363f094cce59dc13d9a6e6243528776fcc3c
* Multi targeting netcoreapp3.1, netstandard2.1 and net5.0 - 8e254b1c3cfbd15298942c4016b4e2483b44b34c
* Implement pluggable storage for activity input and output properties - 332741d54fb64d1d60f2987964f6c2f80d1a7535
* Implement support for HTTP file download for SendHttpRequest and WriteHttpResponse - ebaeabb934d97ed8077d0660fc3ff58247a23f7a
* Implement Blob Storage Workflow Provider - 618f4e837cc09c62e18f4792ef4fc17237df44a3
* Add support for email attachments - 97a9619fad4fd3e8859334de59da2f99f219ec65
* Webhook Management - a8b4c7c3e2d064384d6cb7d91cf99af92a3b5077
* Expose HTTP client creation events Closes #1164 - f50674a6b418ae3ddeb9693e58530babe78f43ea
* Workflow priority queues - 40b53b24a61fe236877af5764b32d3e1368d2d7e

### What's Changed

* **Switch to MIT license** - 76f19a9531c883bccb5de2dec501d6291e9db93e
* **BREAKING** Update OutcomeResult to NOT auto-connect to Done connections - eb378865cce66371aed97da4a5a99a6691e03aef
* Rename ActivityPropertyAttribute to ActivityInputAttribute - 58af0c7ca05b6f246ded634733fdb230bdec1979
* Remove activity state from execution log to prevent large record sizes - ec172d98c1e2fd68b4e8519a6b1087fd14e03308, 7f6c555307ca3157cc898f512167501c223444ae
* Move Dispatch and Execute API endpoints to Workflows resource - c2c9e9cd93e807c73ae581cfa84d1192389ff918
* Increase default code editor height - d1c73d5acdd40bf1a2fa733a5a324b6f00da26d5
* HTTP endpoint handling a single workflow - 8f9431d72a57f3425b0a81f283e95871e3505875
* HTTP endpoint respond with 500 for faulted workflow - ef6f441943d6b4018fbb6d0fe4daf5904ab68763
* Refactor quartz job scheduler into async background service - 8e97f7ae7932ff5f769a2dccf25d9970317cd86b
* Update switch-plugin to support ANY activity type - Closes #1240
* Remove redundant Trigger parameter from workflow selection - cbfb21c4c91e3efcd4e7b83a8cd9f569f6a9e9d8

#### General

* Update WriteHttpResponse to include configurable charset - Fixes #450 - 36061c6802665d95e351f3213f0358dbcbda43aa
* Update WriteHttpResponse activity's ResponseHeaders property's UI hint to MultiLine - a8f6cf37dbf7023fe42878006a42f9b01f7f2828
* Fix race condition between finding bookmarks and indexing workflows - 93e80b52480fb56b31a3bd57a8cccf8be45d1941
* Acquire lock on correlation ID when dispatching new workflows - 97df3b7bc7f951eb94965df6d53adee18679f053
* Fix potential race condition to Quartz scheduler causing duplicate scheduler error - b4f83f43aa9b1c7ec3ad0e7c4ccf26cff6266509
* Fix default value issue with HttpEnpoint - 6e2667716c4438586e342e61a7f3bceadc892e26
* Fix distributed deadlock - 2d12b5c516d9092cd2f7a246b79d5182298c2e74
* Add SendHttpRequest designer plugin - Fixes #1069 - 6e92b6c36ecfd056988bbace759a8a7c684d6122
* Enable content headers for SendHttpRequest - Fixes #1078 - ee898ac6245ca9fe83568736b478fd6eb228182f
* Fix WorkflowLaunchpad correlation locking - 26bf218da68f187397194dc72585b3d5041ba569
* Implement scoped naming for composite activity children - Fixes #1071 - a33cb1bc88bae6c156055b3a307a602a514b4bd8
* Add support for JSON array responses - Fixes #1076 - 5ff1ce7efe8ab4c4ea8c91e22d0f9d0fa32cea9d
* Handle conversion of numeric value to enum and JToken values to strong types - Fixes #1088 - 4f9f56fd976a538b0ad9d3e16aab620d8126e357
* Fix CachingWorkflowRegistry to use cached workflow blueprints - 6e2601ec18a08863df0831db380a00926c66cb12
* State sets default value if get doesn't find it - fe9675715244cd4aefd75cf001d82788046ab801
* Fix double scheduling of activities - 189883ee7f064a08dfab6894556dea3efa88e1ab
* Fix workflow variable access - Fixes #1209 - e427a275ced864daa738102c011414eeec9574c7
* Fix workflow resumption when using ActivityExecuted and WorkflowPassCompleted strategies - Fixes #1182 - c00585241eca96c0b8a8cc8466c5929f3522f569
* Use string instead of Uri to hold request path - See #1070 - a24b1acb0493536a75695be5ad27fa8c7b869ff0
* Fix Redirect behavior - See https://stackoverflow.com/questions/68217426/elsa-workflows-redirect-http-action-behavior
* Normalize signal names (ignore case) - 88bc022140ac9fe15cf73626089d5abc4c228b15
* Fix Finish output serialization recursion - 7452d16ebe1d626ba14329b0870d43ed752fdc82
* Fix Join activity not blocking when using WaitAll mode - f65cf6b35e663e9e48c962fc341020e8150ebe45
* Add client plugin for RunWorkflow to support dynamic outcomes - 505c00addfc22d4781450353d7d6c48958b0f1c1
* Fix that workflows blocked on a signal could not be resumed with just correlation ID - 0850f0ac486be7828d96ab4906c3abf8045382c2
* Fix StartupTask DI registration - 5ae4609165625184deedb989a1b0e1c8728e3502
* Fix HttpEndpointMiddleware to return 404 when workflow not found - Fixes #1253 - 2635115766189e83f66dc1b05a49b969887e1f73
* Fix CompositeActivityBlueprint initialization - e3120e9329e41ec4bf0d0550a1745026265c88b4
* Reset IsLatest also for not published workflows on PublishAsync - a42df38f6e27b32b6456f89c330d19028d8d5a6b

#### Workflow Expressions

* Add JS functions to access activity properties - 7f420f4f42e8152e7763cc49df9f6c1bb9fbfae6
* Increase javascript intellisense coverage - d033b56d29ad1fe5905bbf33f1f2e2d797da7111
* Fix liquid engine to handle JObject and JValue resolution - d195dfc070f0541416af55285347302ca5a05ee6

#### MongoDB

* **BREAKING**: Update MongoDB provider with support for object derivatives - Fixes #552 - 45b0f84feec70299f6b6574eb347077f4eda300e 
* Fix `Variables` serialization for MongoDB provider - da8d12b9e4ced98b77efc4c7dc8cd943270e5093
* Fix Type serialization for MongoDB provider - abc4e5389c0acc5ac7b8fc3b84e7d752effe23bc

#### Entity Framework Core

* Fix MySql does not support the EF Core concept of schemas - 34556e967070bfad1db7ce80abffd9d76358c3f5
* Fix EFCore.BulkExtensions does not support MySQL - 3b1ccd4f6a15dc45dc36644e951896a539a259e5

#### Dashboard / Designer

* Apply auto-scroll on exception details - f5b688a3dcf7cc9be44a4a9da9f4431665488752
* Fix issue with SwitchHandler and Switch display on workflow blueprint viewer when using expressions that cannot be evaluated - 68e5be0159f5a4ce27b0639597befcc1265470ed
* Uncheck "check all" after bulk delete - Fixes #1057 - 88a5c61a164d8edeb3660ea504c37a4c64f0e9a7
* Improve workflow journal details - 7958ecf49a7a924865c86be3a6e0c559973f2340
* Optimize workflow definitions loading - 6f8990def29f7958c2fb1823b5ad9d2a602c3003
* Fix JS error when using workflow instance list view in standalone mode - Fixes #1235 - dd55189bedfc94d1979477b18648233670fefe16
* Prevent display text from pushing settings button out of view - Fixes #1227 - dbb91a816a534fd5db848ecfb63f33e9803da8c0

<hr /><em>This discussion was created from the release <a href='https://github.com/elsa-workflows/elsa-core/releases/tag/2.1'>Elsa 2.1 ✨</a>.</em>

**Comment by PeeraJ** [comment]
Thanks for your hard work! Good job guys! keep it up!🎉🎉🎉
